### PR TITLE
Support coercion from Owned(Child) to Owned(Parent)

### PR DESCRIPTION
**Pull request by mppf**
This PR enables value coercions from Owned(Child) to Owned(Parent). Coercions in a reference context, such as with a `const ref` argument are not currently supported. Additionally coercing a class (not owned) to a parent class type for a `const ref` argument results in backend C compiler warnings from pointer type mismatches. This PR creates a new error message in resolution for that case which also causes errors for passing an Owned that needs coercion to a const ref argument.

- [x] full local testing
- [x] add Owned example from #8489
- [x] futures testing
- [x] baseline testing

Reviewed by @daviditen - thanks!

### CODE REVIEW

File: compiler/resolution/wrappers.cpp [hunk 1]
```diff
@@ -1592,6 +1592,14 @@ static void addArgCoercion(FnSymbol*  fn,
   }
 
   resolveCall(castMove);
+
+  IntentTag concreteIntent = concreteIntentForArg(formal);
+  bool formalIsRef = formal->isRef() || (concreteIntent & INTENT_REF);
+
+  if (!castTemp->isRef() && formalIsRef) {
+    USR_FATAL_CONT(call, "Cannot coerce argument for const ref intent");
+  }
```
Comment (mppf) [inline compiler/resolution/wrappers.cpp]: I think it'd be better to fold this into a checking function that includes/replaces calls to argumentCanModifyActual/issueErrorValueTemporaryToRef

**Comment by mppf** [comment]
@daviditen - do you think the error in classes/diten/passSubClassByRef is good enough to un-future it? I mention this question in the last comment in issue #8489. AFAIK we don't currently choose candidates based upon the mutability of the actual and I think it'd be a big change to do so. I suspect the unresolved call error was there more as an example of some sort of error, though.

File: compiler/resolution/wrappers.cpp [hunk 2]
```diff
@@ -1435,8 +1431,33 @@ static bool argumentCanModifyActual(ArgSymbol* formal) {
...
+    // OK with a 'const ref' but right now there are errors at C
+    // compilation time if this error is left out.
+    // Additionally, if a new value is created for this kind of
+    // coercion, it disrupts the desired semantics (a value passed
+    // by const ref could be modified during the call & the change
+    // visible).
+    bool formalIsRef = formal->isRef() || (intent & INTENT_REF);
+
+    if (formalIsRef) {
+      USR_FATAL_CONT(call, "Cannot coerce argument for const ref intent");
```
Comment (daviditen) [inline compiler/resolution/wrappers.cpp]: It would be nice to print the formal name here too.

File: modules/packages/OwnedObject.chpl [hunk 3]
```diff
@@ -204,7 +204,7 @@ module OwnedObject {
 
   pragma "no doc"
   inline proc _cast(type t, in x) where t:Owned && x:Owned && x.t:t.t {
-    var ret = new Owned(x.release());
+    var ret = new Owned(x.release():t.t);
```
Comment (daviditen) [inline modules/packages/OwnedObject.chpl]: A comment here explaining why the cast is needed would be helpful.  It took a minute staring at it to figure it out.

File: test/classes/diten/passSubClassByRef.bad [hunk 4]
```diff
@@ -1,7 +1 @@
-passSubClassByRef.chpl:6: internal error: CHEnnnn chpl version mmmm
-
-Internal errors indicate a bug in the Chapel compiler ("It's us, not you"),
-and we're sorry for the hassle.  We would appreciate your reporting this bug -- 
-please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
-the filename + line number above may be useful in working around the issue.
-
+passSubClassByRef.chpl:6: error: value from coercion passed to ref formal 'obj'
```
Comment (daviditen) [inline test/classes/diten/passSubClassByRef.bad]: This works for me to de-futurize.

**Pull request merged**